include/cudaq/Optimizer/Transforms/Passes.td

```diff
@@ -214,6 +215,49 @@ def DecompositionPass: Pass<"decomposition", "mlir::ModuleOp"> {
   ];
 }
 
+def DependencyAnalysis : Pass<"dep-analysis", "mlir::ModuleOp"> {
+  let summary = "Maps qubits and reorders operations based on dependency graph.";
+  let description = [{
+    A dependency graph is a Directed Acyclic Graph (DAG) where each node
+    represents an operation, and each edge represents a "depends on" relation
+    between that operation and another operation. For example, in the following
+    snippet the `x` operation depends on the `h` operation because it is applied
+    to the same qubit (`q`), so the `h` operation must happen before the `x`
+    operation:
+    ```c++
+      cudaq::qubit q;
+      x(q);
+      h(q);
+    ```
+
+    Once a dependency graph is created, it is then scheduled, assigning each
+    operation a virtual cycle. Operations that don't depend on each other may
+    be scheduled at the same cycle. However, an operation that depends on a
+    second operation must be scheduled after the second operation. The
+    scheduling algorithm tries to pack operations as densely as possible,
+    minimizing the number of cycles between operations.
+
+    From this dependency graph, we can calculate the lifetime of a qubit: from
+    the cycle in which it is first used through the cycle in which it is last
+    used. If two virtual qubits have non-overlapping lifetimes, they can be
+    assigned to the same physical qubit, as every virtual qubit is assumed to be
+    fully reset before release. Failure to fully reset virtual qubits before
+    release is undefinied behavior, and will likely lead to incorrect output
+    when running DependencyAnalysis.
```

Comment:
Noting for future: we need to highlight this caveat in any eventual user-facing documentation.

Reply:
I think we may need to auto-insert the `reset` ops. I don't know that this happens, per se, but it could be an issue.
```
  // gates
  mz qubit
  return_wire qubit
```
converted to
```
  // ...
  mz qubit
  // reuse qubit now, replacing qubit2
  // error? qubit needs to be set to |0>
  // ... [qubit2\qubit] ... 
  return_wire qubit
```

Reply:
Will create a separate issue for this one (or add to a new grouped issue).